[BUG] Gemini PartsMapper throws MalformedJsonException when using VertexAiGeminiChatModel

**Describe the bug**
The `com.google.gson.stream.MalformedJsonException` exception is thrown during `ToolExecutionResultMessage` handling when using the `VertexAiGeminiChatModel` via `AiServices`. This occurs when the tool response contains escaped quotes (e.g., `\"`).

**Log and Stack trace**
```java
java.lang.RuntimeException: com.google.protobuf.InvalidProtocolBufferException: com.google.gson.stream.MalformedJsonException: Unterminated object at line 1 column 28 path $.result
See https://github.com/google/gson/blob/main/Troubleshooting.md#malformed-json
	at dev.langchain4j.model.vertexai.gemini.PartsMapper.map(PartsMapper.java:137)
	at dev.langchain4j.model.vertexai.gemini.ContentsMapper.createContent(ContentsMapper.java:85)
	at dev.langchain4j.model.vertexai.gemini.ContentsMapper.splitInstructionAndContent(ContentsMapper.java:45)
	at dev.langchain4j.model.vertexai.gemini.VertexAiGeminiChatModel.generate(VertexAiGeminiChatModel.java:425)
	at dev.langchain4j.model.vertexai.gemini.VertexAiGeminiChatModel.chat(VertexAiGeminiChatModel.java:373)
	at dev.langchain4j.service.tool.ToolService.executeInferenceAndToolsLoop(ToolService.java:308)
	at dev.langchain4j.service.DefaultAiServices$1.invoke(DefaultAiServices.java:392)
	at dev.langchain4j.service.DefaultAiServices$1.invoke(DefaultAiServices.java:236)
	at com.karsta26.chatai.$Proxy2.chat(Unknown Source)
	at com.karsta26.chatai.GeminiChat.main(GeminiChat.java:34)
Caused by: com.google.protobuf.InvalidProtocolBufferException: com.google.gson.stream.MalformedJsonException: Unterminated object at line 1 column 28 path $.result
See https://github.com/google/gson/blob/main/Troubleshooting.md#malformed-json
	at com.google.protobuf.util.JsonFormat$ParserImpl.merge(JsonFormat.java:1320)
	at com.google.protobuf.util.JsonFormat$Parser.merge(JsonFormat.java:449)
	at dev.langchain4j.model.vertexai.gemini.PartsMapper.map(PartsMapper.java:135)
	... 9 more
Caused by: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unterminated object at line 1 column 28 path $.result
See https://github.com/google/gson/blob/main/Troubleshooting.md#malformed-json
	at com.google.gson.internal.Streams.parse(Streams.java:58)
	at com.google.gson.JsonParser.parseReader(JsonParser.java:144)
	at com.google.protobuf.util.JsonFormat$ParserImpl.merge(JsonFormat.java:1317)
	... 11 more
Caused by: com.google.gson.stream.MalformedJsonException: Unterminated object at line 1 column 28 path $.result
See https://github.com/google/gson/blob/main/Troubleshooting.md#malformed-json
	at com.google.gson.stream.JsonReader.syntaxError(JsonReader.java:1817)
	at com.google.gson.stream.JsonReader.doPeek(JsonReader.java:616)
	at com.google.gson.stream.JsonReader.hasNext(JsonReader.java:538)
	at com.google.gson.internal.bind.JsonElementTypeAdapter.read(JsonElementTypeAdapter.java:95)
	at com.google.gson.internal.bind.JsonElementTypeAdapter.read(JsonElementTypeAdapter.java:35)
	at com.google.gson.internal.Streams.parse(Streams.java:46)
	... 13 more
```

**To Reproduce**
```java
public class GeminiChat {

    static class Calculator {
        @Tool("Adds two given numbers")
        String add(double a, double b) {
            System.out.println("Adding " + a + " and " + b);
            return "The result is \"" + (a + b) + "\".";
        }
    }

    interface Assistant {
        String chat(String userMessage);
    }

    public static void main(String[] args) {
        ChatModel model = VertexAiGeminiChatModel.builder()
                .project(System.getenv("PROJECT_ID"))
                .location("us-central1")
                .modelName("gemini-2.5-flash")
                .build();

        Assistant assistant = AiServices.builder(Assistant.class)
                .chatModel(model)
                .tools(new Calculator())
                .build();

        String response = assistant.chat("How much is 74589613588 + 4786521789?");

        System.out.println(response);
    }
}
```

**Expected behavior**
No exception is thrown, the tool returns its result.

**Please complete the following information:**
- LangChain4j version: 1.9.1
- LLM(s) used: gemini-2.5-flash 
- Java version: 21
- Spring Boot version (if applicable): not used

**Additional context**

It looks like this code was not fully finished (there is a `TODO` in the code):

<img width="923" height="725" alt="Image" src="https://github.com/user-attachments/assets/fda1c2a9-352f-454d-8ac7-ccdf0a218e1c" />

--- a/langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/PartsMapper.java
+++ b/langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/PartsMapper.java
@@ -1,9 +1,16 @@
 package dev.langchain4j.model.vertexai.gemini;
 
+import static com.google.cloud.vertexai.generativeai.PartMaker.fromMimeTypeAndData;
+import static dev.langchain4j.internal.Exceptions.illegalArgument;
+import static dev.langchain4j.internal.Utils.getOrDefault;
+import static dev.langchain4j.internal.Utils.readBytes;
+import static java.util.Collections.singletonList;
+
 import com.google.cloud.vertexai.api.FunctionResponse;
 import com.google.cloud.vertexai.api.Part;
 import com.google.protobuf.InvalidProtocolBufferException;
 import com.google.protobuf.Struct;
+import com.google.protobuf.Value;
 import com.google.protobuf.util.JsonFormat;
 import dev.langchain4j.data.audio.Audio;
 import dev.langchain4j.data.image.Image;
@@ -20,22 +27,12 @@ import dev.langchain4j.data.message.UserMessage;
 import dev.langchain4j.data.message.VideoContent;
 import dev.langchain4j.data.pdf.PdfFile;
 import dev.langchain4j.data.video.Video;
-import dev.langchain4j.internal.CustomMimeTypesFileTypeDetector;
-
 import java.net.URI;
+import java.util.ArrayList;
 import java.util.Base64;
 import java.util.HashMap;
-import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
-
-import static com.google.cloud.vertexai.generativeai.PartMaker.fromMimeTypeAndData;
-import static dev.langchain4j.internal.Exceptions.illegalArgument;
-import static dev.langchain4j.internal.Utils.getOrDefault;
-import static dev.langchain4j.internal.Utils.quoted;
-import static dev.langchain4j.internal.Utils.readBytes;
-import static java.util.Collections.singletonList;
-import static java.util.stream.Collectors.toList;
 
 class PartsMapper {
 
@@ -86,91 +83,62 @@ class PartsMapper {
     }
 
     static List<Part> map(ChatMessage message) {
-        if (message instanceof AiMessage) {
-            AiMessage aiMessage = (AiMessage) message;
+        if (message instanceof AiMessage aiMessage) {
 
             List<Part> parts = new ArrayList<>();
 
             if (aiMessage.text() != null && !aiMessage.text().isEmpty()) {
-                parts.add(Part.newBuilder()
-                    .setText(aiMessage.text())
-                    .build());
+                parts.add(Part.newBuilder().setText(aiMessage.text()).build());
             }
 
             if (aiMessage.hasToolExecutionRequests()) {
                 List<Part> fnCallReqParts = aiMessage.toolExecutionRequests().stream()
-                    .map(FunctionCallHelper::fromToolExecutionRequest)
-                    .map(fnCall -> Part.newBuilder()
-                        .setFunctionCall(fnCall)
-                        .build())
-                    .collect(toList());
+                        .map(FunctionCallHelper::fromToolExecutionRequest)
+                        .map(fnCall -> Part.newBuilder().setFunctionCall(fnCall).build())
+                        .toList();
 
                 parts.addAll(fnCallReqParts);
             }
 
             return parts;
-        } else if (message instanceof UserMessage) {
-            return ((UserMessage) message).contents().stream()
-                .map(PartsMapper::map)
-                .collect(toList());
-        } else if (message instanceof SystemMessage) {
-            return singletonList(Part.newBuilder()
-                .setText(((SystemMessage) message).text())
-                .build());
-        } else if (message instanceof ToolExecutionResultMessage) {
-            ToolExecutionResultMessage toolExecutionResultMessage = (ToolExecutionResultMessage) message;
+        } else if (message instanceof UserMessage userMessage) {
+            return userMessage.contents().stream().map(PartsMapper::map).toList();
+        } else if (message instanceof SystemMessage systemMessage) {
+            return singletonList(
+                    Part.newBuilder().setText((systemMessage).text()).build());
+        } else if (message instanceof ToolExecutionResultMessage toolExecutionResultMessage) {
             String functionResponseText = toolExecutionResultMessage.text();
-
-            Struct.Builder structBuilder = Struct.newBuilder();
-            try {
-                JsonFormat.parser().merge(functionResponseText, structBuilder);
-            } catch (InvalidProtocolBufferException e) {
-                // TODO do a proper fix
-                String functionResponseTextAsMap = "{\"result\":" + functionResponseText + "}";
-                try {
-                    JsonFormat.parser().merge(functionResponseTextAsMap, structBuilder);
-                } catch (InvalidProtocolBufferException e2) {
-                    String functionResponseTextWithQuotesAsMap = "{\"result\":" + quoted(functionResponseText) + "}";
-                    try {
-                        JsonFormat.parser().merge(functionResponseTextWithQuotesAsMap, structBuilder);
-                    } catch (InvalidProtocolBufferException e3) {
-                        throw new RuntimeException(e3);
-                    }
-                }
-            }
-            Struct responseStruct = structBuilder.build();
+            Struct responseStruct = parseFunctionResponseToStruct(functionResponseText);
 
             return singletonList(Part.newBuilder()
-                .setFunctionResponse(FunctionResponse.newBuilder()
-                    .setName(toolExecutionResultMessage.toolName())
-                    .setResponse(responseStruct)
-                    .build())
-                .build());
+                    .setFunctionResponse(FunctionResponse.newBuilder()
+                            .setName(toolExecutionResultMessage.toolName())
+                            .setResponse(responseStruct)
+                            .build())
+                    .build());
         } else {
             throw illegalArgument(message.type() + " message is not supported by Gemini");
         }
     }
 
     private static Part map(Content content) {
-        if (content instanceof TextContent) {
-            return map((TextContent) content);
-        } else if (content instanceof ImageContent) {
-            return map((ImageContent) content);
-        } else if (content instanceof AudioContent) {
-            return map((AudioContent) content);
-        } else if (content instanceof VideoContent) {
-            return map((VideoContent) content);
-        } else if (content instanceof PdfFileContent) {
-            return map((PdfFileContent) content);
+        if (content instanceof TextContent text) {
+            return map(text);
+        } else if (content instanceof ImageContent image) {
+            return map(image);
+        } else if (content instanceof AudioContent audio) {
+            return map(audio);
+        } else if (content instanceof VideoContent video) {
+            return map(video);
+        } else if (content instanceof PdfFileContent pdf) {
+            return map(pdf);
         } else {
             throw illegalArgument("Unknown content type: " + content);
         }
     }
 
     private static Part map(TextContent content) {
-        return Part.newBuilder()
-                .setText(content.text())
-                .build();
+        return Part.newBuilder().setText(content.text()).build();
     }
 
     static Part map(ImageContent content) {
@@ -215,5 +183,30 @@ class PartsMapper {
             }
         }
         throw illegalArgument("Unable to detect the MIME type of '%s'. Please provide it explicitly.", url);
+    }
+
+    private static Struct parseFunctionResponseToStruct(String functionResponseText) {
+        try {
+            Struct.Builder structBuilder = Struct.newBuilder();
+            JsonFormat.parser().merge(functionResponseText, structBuilder);
+            return structBuilder.build();
+        } catch (InvalidProtocolBufferException e) {
+            // Not a JSON object
+        }
+
+        try {
+            String wrappedJson = "{\"result\":" + functionResponseText + "}";
+            Struct.Builder structBuilder = Struct.newBuilder();
+            JsonFormat.parser().merge(wrappedJson, structBuilder);
+            return structBuilder.build();
+        } catch (InvalidProtocolBufferException e) {
+            // Not a valid JSON
+        }
+
+        return Struct.newBuilder()
+                .putFields(
+                        "result",
+                        Value.newBuilder().setStringValue(functionResponseText).build())
+                .build();
     }
 }

--- a/langchain4j-vertex-ai-gemini/src/test/java/dev/langchain4j/model/vertexai/gemini/ContentsMapperTest.java
+++ b/langchain4j-vertex-ai-gemini/src/test/java/dev/langchain4j/model/vertexai/gemini/ContentsMapperTest.java
@@ -312,4 +312,529 @@ class ContentsMapperTest {
                         .getName())
                 .isEqualTo("fetchData");
     }
+
+    @Test
+    void should_handle_tool_result_as_json_object() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getData", "{\"name\": \"John\", \"age\": 30}"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("name").getStringValue()).isEqualTo("John");
+        assertThat(response.getFieldsMap().get("age").getNumberValue()).isEqualTo(30.0);
+    }
+
+    @Test
+    void should_handle_tool_result_as_json_array() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getList", "[1, 2, 3]"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getListValue().getValuesCount())
+                .isEqualTo(3);
+        assertThat(response.getFieldsMap()
+                        .get("result")
+                        .getListValue()
+                        .getValues(0)
+                        .getNumberValue())
+                .isEqualTo(1.0);
+        assertThat(response.getFieldsMap()
+                        .get("result")
+                        .getListValue()
+                        .getValues(1)
+                        .getNumberValue())
+                .isEqualTo(2.0);
+    }
+
+    @Test
+    void should_handle_tool_result_as_json_number() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "calculate", "42"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getNumberValue()).isEqualTo(42.0);
+    }
+
+    @Test
+    void should_handle_tool_result_as_json_boolean() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "checkStatus", "true"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getBoolValue()).isTrue();
+    }
+
+    @Test
+    void should_handle_tool_result_as_json_string() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getMessage", "\"hello world\""));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getStringValue()).isEqualTo("hello world");
+    }
+
+    @Test
+    void should_handle_tool_result_as_plain_text() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "describe", "This is plain text output"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getStringValue()).isEqualTo("This is plain text output");
+    }
+
+    @Test
+    void should_handle_tool_result_as_plain_text_with_embedded_quotes() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "add", "The result is \"79376135377\"."));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getStringValue()).isEqualTo("The result is \"79376135377\".");
+    }
+
+    @Test
+    void should_handle_tool_result_as_json_null() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "findItem", "null"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").hasNullValue()).isTrue();
+    }
+
+    @Test
+    void should_handle_tool_result_as_nested_json_object() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(
+                null, "getNestedData", "{\"user\": {\"name\": \"Alice\", \"address\": {\"city\": \"Paris\"}}}"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        Struct user = response.getFieldsMap().get("user").getStructValue();
+        assertThat(user.getFieldsMap().get("name").getStringValue()).isEqualTo("Alice");
+        Struct address = user.getFieldsMap().get("address").getStructValue();
+        assertThat(address.getFieldsMap().get("city").getStringValue()).isEqualTo("Paris");
+    }
+
+    @Test
+    void should_handle_tool_result_as_negative_number() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getDifference", "-42"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("result").getNumberValue()).isEqualTo(-42.0);
+    }
+
+    @Test
+    void should_handle_tool_result_as_floating_point_number() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getFloat", "5.64659"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("result").getNumberValue()).isEqualTo(5.64659);
+    }
+
+    @Test
+    void should_handle_tool_result_as_empty_json_array() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getEmptyList", "[]"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("result").getListValue().getValuesCount())
+                .isZero();
+    }
+
+    @Test
+    void should_handle_tool_result_as_array_of_objects() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getUsers", "[{\"name\": \"Alice\"}, {\"name\": \"Bob\"}]"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("result").getListValue().getValuesCount())
+                .isEqualTo(2);
+        assertThat(response.getFieldsMap()
+                        .get("result")
+                        .getListValue()
+                        .getValues(0)
+                        .getStructValue()
+                        .getFieldsMap()
+                        .get("name")
+                        .getStringValue())
+                .isEqualTo("Alice");
+    }
+
+    @Test
+    void should_handle_tool_result_with_special_characters() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getSpecial", "{\"text\": \"Line1\\nLine2\\tTabbed\"}"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("text").getStringValue()).isEqualTo("Line1\nLine2\tTabbed");
+    }
+
+    @Test
+    void should_handle_tool_result_with_unicode_characters() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getEmoji", "{\"message\": \"Hello 👋 World 🌍\"}"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("message").getStringValue()).isEqualTo("Hello 👋 World 🌍");
+    }
+
+    @Test
+    void should_handle_tool_result_as_scientific_notation_number() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getScientific", "1.5e10"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("result").getNumberValue()).isEqualTo(1.5e10);
+    }
+
+    @Test
+    void should_handle_tool_result_with_json_containing_backslashes() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getPath", "{\"path\": \"C:\\\\Users\\\\test\"}"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("path").getStringValue()).isEqualTo("C:\\Users\\test");
+    }
+
+    @Test
+    void should_handle_tool_result_as_mixed_type_array() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getMixed", "[1, \"two\", true, null]"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("result").getListValue().getValuesCount())
+                .isEqualTo(4);
+        assertThat(response.getFieldsMap()
+                        .get("result")
+                        .getListValue()
+                        .getValues(0)
+                        .getNumberValue())
+                .isEqualTo(1.0);
+        assertThat(response.getFieldsMap()
+                        .get("result")
+                        .getListValue()
+                        .getValues(1)
+                        .getStringValue())
+                .isEqualTo("two");
+        assertThat(response.getFieldsMap()
+                        .get("result")
+                        .getListValue()
+                        .getValues(2)
+                        .getBoolValue())
+                .isTrue();
+        assertThat(response.getFieldsMap()
+                        .get("result")
+                        .getListValue()
+                        .getValues(3)
+                        .hasNullValue())
+                .isTrue();
+    }
+
+    @Test
+    void should_handle_tool_result_as_empty_json_object() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getEmptyObject", "{}"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsCount()).isZero();
+    }
+
+    @Test
+    void should_handle_tool_result_text_starting_with_curly_brace_but_invalid_json() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getInvalid", "{not valid json}"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getStringValue()).isEqualTo("{not valid json}");
+    }
+
+    @Test
+    void should_handle_tool_result_text_starting_with_bracket_but_invalid_json() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getInvalidArray", "[not, valid, json"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getStringValue()).isEqualTo("[not, valid, json");
+    }
+
+    @Test
+    void should_handle_tool_result_with_html_content() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(null, "getHtml", "<html><body>Hello</body></html>"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getStringValue()).isEqualTo("<html><body>Hello</body></html>");
+    }
+
+    @Test
+    void should_handle_tool_result_with_xml_content() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(
+                null, "getXml", "<?xml version=\"1.0\"?><root><item>value</item></root>"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap()).containsKey("result");
+        assertThat(response.getFieldsMap().get("result").getStringValue())
+                .isEqualTo("<?xml version=\"1.0\"?><root><item>value</item></root>");
+    }
+
+    @Test
+    void should_handle_tool_result_with_json_containing_url() {
+        // given
+        List<ChatMessage> msgs = new ArrayList<>();
+        msgs.add(ToolExecutionResultMessage.from(
+                null, "getUrl", "{\"url\": \"https://example.com/path?query=value&other=123\"}"));
+
+        // when
+        ContentsMapper.InstructionAndContent instructionAndContent = splitInstructionAndContent(msgs);
+
+        // then
+        Struct response = instructionAndContent
+                .contents
+                .get(0)
+                .getParts(0)
+                .getFunctionResponse()
+                .getResponse();
+        assertThat(response.getFieldsMap().get("url").getStringValue())
+                .isEqualTo("https://example.com/path?query=value&other=123");
+    }
 }

